### PR TITLE
feat(core): add dvcr-cleaner to dvcr image

### PR DESCRIPTION
**Pull request by hardcoretime**
## Description

dvcr-cleaner can:
 - list images
 - delete images
 - run `garbage-collect`
## Why do we need it, and what problem does it solve?

When a virtual image resource is deleted from the API, it is not deleted from the registry.  
`dvcr-cleaner` is a temporary tool for cleaning the dvcr.   
## What is the expected result?

`dvcr-cleaner` works properly in a running dvcr container.
## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+		if len(args) != 0 {
+			image := args[0]
+			info, err := os.Stat(fmt.Sprintf("%s/vi/%s/%s", ImageRepoDir, NamespaceFlag, image))
+			if errors.Is(err, os.ErrNotExist) {
+				return fmt.Errorf("virtual image %q not found in %q namespace", image, NamespaceFlag)
+			}
+			if err != nil {
+				return fmt.Errorf("cannot get virtual image %q in %q namespace: %w", image, NamespaceFlag, err)
+			}
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go:65]: bad pattern. user
```suggestion
			if err != nil {
			    if errors.Is(err, os.ErrNotExist) {
				    return fmt.Errorf("virtual image %q not found in %q namespace", image, NamespaceFlag)
			    }
				return fmt.Errorf("cannot get virtual image %q in %q namespace: %w", image, NamespaceFlag, err)
			}
```

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+			fmt.Println(info.Name())
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go:66]: use fmt.Fprintln with os.Stdout.
And could you make a table?

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		if AllImagesFlag {
+			images, err := ListVirtualImages(NamespaceFlag)
+			if err != nil {
+				return err
+			}
+			fmt.Println(strings.Join(images, "\n"))
+			return nil
+		}
+
+		if AllNamespacesFlag {
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go:79]: I think this flag should have higher priority than AllImagesFlag.

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+func ListVirtualImages(namespace string) ([]string, error) {
+	nsPath := fmt.Sprintf("%s/vi/%s", ImageRepoDir, namespace)
+	files, err := os.ReadDir(nsPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get list of all images in %q namespace: %w", namespace, err)
+	}
+	images := make([]string, 0, len(files))
+	for _, f := range files {
+		images = append(images, fmt.Sprintf("%s/%s", namespace, f.Name()))
+	}
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go:155]:
```suggestion
	images := make([]string, len(files))
	for i, f := range files {
		images[i] = fmt.Sprintf("%s/%s", namespace, f.Name())
	}
```

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	return images, nil
+}
+
+func init() {
+	LsCmd.AddCommand(lsCviCmd, lsViCmd)
+	lsViCmd.Flags().StringVarP(&NamespaceFlag, "namespace", "n", "default", "namespace of virtual images")
+	lsViCmd.Flags().BoolVar(&AllImagesFlag, "all", false, "list of all virtual images")
+	lsViCmd.Flags().BoolVar(&AllNamespacesFlag, "all-namespaces", false, "list of all virtual images in all namespaces")
+	lsViCmd.MarkFlagsMutuallyExclusive("all", "all-namespaces")
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go:164]: why do you need in all-namespaces this?

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+	Args: cobra.MatchAll(
+		cobra.MaximumNArgs(1),
+		cobra.OnlyValidArgs,
+	),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) == 0 && !AllImagesFlag && !AllNamespacesFlag {
+			return cmd.Help()
+		}
+
+		if len(args) != 0 && AllImagesFlag {
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go:53]:
```suggestion
		if len(args) != 0 && (AllImagesFlag || AllNamespacesFlag) {
```

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/delete.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/spf13/cobra"
+)
+
+var DeleteCmd = &cobra.Command{
+	Use:   "delete",
+	Short: "Delete virtual images",
+	Args:  cobra.MatchAll(cobra.ExactArgs(1), cobra.OnlyValidArgs),
+}
+
+var deleteViCmd = &cobra.Command{
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/delete.go:34]: looks almost identical to deleteCviCmd. You can put the logic into a separate function

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+	ImageRepoDir      = "/var/lib/registry/docker/registry/v2/repositories"
+)
+
+var LsCmd = &cobra.Command{
+	Use:   "ls",
+	Short: "List of virtual images",
+	Args:  cobra.MatchAll(cobra.ExactArgs(1), cobra.OnlyValidArgs),
+}
+
+var lsViCmd = &cobra.Command{
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go:41]: looks almost identical to lsCviCmd. You can put the logic into a separate function

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/gc.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+	Short: "run `garbage-collect` deletes cache data and layers not referenced by any manifests",
+	Args:  cobra.OnlyValidArgs,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		err := Confirm()
+		if errors.Is(err, promptui.ErrAbort) {
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("confirm is failed: %w", err)
+		}
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/gc.go:46]: bad pattern

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/gc.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+		IsConfirm: true,
+	}
+
+	_, err := prompt.Run()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/gc.go:78]:
```suggestion
func Confirm() (bool, error) {
	prompt := promptui.Prompt{
		Label:     "Confirm",
		IsConfirm: true,
	}

	_, err := prompt.Run()
	if err != nil {
	    if errors.Is(err, promptui.ErrAbort) {
			return false, nil
		}
		return false, err
	}

	return true, nil
}
```

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go (hunk 3 above)
Comment (hardcoretime) [reply images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go]: `all` and `all-namespaces` cannot be used at the same time.
https://github.com/deckhouse/virtualization/blob/57fb97971074beab9d12b9c932ab77e4468fd611/images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go#L197

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go (hunk 5 above)
Comment (hardcoretime) [reply images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go:197]: If we need to find images in all namespaces, but we don't have a list of namespaces, or some namespaces no longer exist in the cluster.

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/delete.go [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+			return err
+		}
+
+		return nil
+	},
+	SilenceUsage:  true,
+	SilenceErrors: true,
+}
+
+func DeleteImage(imageType, imgsDir string, cmd *cobra.Command, args []string) error {
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/delete.go:75]: don't used imagetype

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+			imgs, err := os.ReadDir(imgsDir)
+			if err != nil {
+				return fmt.Errorf("cannot get the list of all `ClusterVirtualImages`: %w", err)
+			}
+			w := tabwriter.NewWriter(os.Stdout, 1, 1, 1, ' ', 0)
+			fmt.Fprintln(w, "Name\t")
+			for _, img := range imgs {
+				fmt.Fprintf(w, "%s\t\n", img.Name())
+			}
+			w.Flush()
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go:145]: same code

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+				return fmt.Errorf("cannot get the `%s` %q in the %q namespace: %w", imgType, imgName, NamespaceFlag, err)
+			}
+		case virtv2.ClusterVirtualImageKind:
+			path := fmt.Sprintf("%s/%s", imgsDir, imgName)
+			fileInfo, err = os.Stat(path)
+			if err != nil {
+				if errors.Is(err, os.ErrNotExist) {
+					return fmt.Errorf("the `%s` %q is not found", imgType, imgName)
+				}
+				return fmt.Errorf("cannot get the `%s` %q: %w", imgType, imgName, err)
```
Comment (yaroslavborbat) [inline images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go:110]: same code

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]
need rebase

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go (hunk 13 above)
Comment (yaroslavborbat) [reply images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go:110]: but ok

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go (hunk 12 above)
Comment (yaroslavborbat) [reply images/dvcr-artifact/cmd/dvcr-cleaner/cmd/ls.go:145]: but ok

File: images/dvcr-artifact/cmd/dvcr-cleaner/cmd/delete.go (hunk 11 above)
Comment (hardcoretime) [reply images/dvcr-artifact/cmd/dvcr-cleaner/cmd/delete.go:75]: it should be used in `removeImageDir` func, thanks

**Review by yaroslavborbat — APPROVED** [review]

**Pull request merged**